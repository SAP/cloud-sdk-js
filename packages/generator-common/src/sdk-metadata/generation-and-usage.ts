import { InstructionWithText } from './sdk-metadata-types';

export function getGenerationSteps(
  installCommand: string,
  generateCommand: string,
  generationDocLink: string
): InstructionWithText {
  return {
    instructions: `<ul>
<li>Download the API specification to your local computer.</li>
<li>Install the generator <code>${installCommand}</code></li>
<li>Execute the generator <code>${generateCommand}</code></li>
</ul>
The steps above will generate a basic TypeScript client. For additional options like transpiling to JavaScript etc. visit our <a href="${generationDocLink}">documentation.</a>`,
    text: 'Follow the following generation steps to generate the client.'
  };
}

export const apiSpecificUsageText = 'Find a usage example for this API below.';

export const genericUsageText =
  'Find a generic example on how to execute requests with the SAP Cloud SDK below.';
