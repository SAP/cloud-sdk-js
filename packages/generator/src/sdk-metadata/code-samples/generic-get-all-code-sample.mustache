import { {{{EntityName}}} } from '{{{PackageName}}}';

const resultPromise = {{{EntityName}}}.requestBuilder().getAll().top(5).execute({ destinationName:'myDestinationName' });
