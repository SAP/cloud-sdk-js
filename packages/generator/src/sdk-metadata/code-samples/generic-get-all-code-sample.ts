import { codeBlock } from '@sap-cloud-sdk/util';

export function genericGetAllCodeSample(
  entityName: string,
  packageName: string
): string {
  return codeBlock`
import { ${entityName} } from '${packageName}';

const resultPromise = ${entityName}.requestBuilder().getAll().top(5).execute({ destinationName:'myDestinationName' });

`;
}
