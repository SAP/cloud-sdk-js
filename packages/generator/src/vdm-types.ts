import { PathLike } from 'fs';
import { ODataVersion } from '@sap-cloud-sdk/util';
/**
 * @internal
 */
export interface VdmServicePackageMetaData {
  /**
   * @internal
   */
  oDataVersion: ODataVersion;
  /**
   * @internal
   */
  namespaces: string[];
  /**
   * @internal
   */
  originalFileName: string;
  /**
   * @internal
   */
  servicePath: string;
  /**
   * @internal
   */
  directoryName: string;
  /**
   * @internal
   */
  npmPackageName: string;
  /**
   * @internal
   */
  speakingModuleName: string;
  /**
   * @internal
   */
  apiBusinessHubMetadata?: ApiBusinessHubMetadata | undefined;
  /**
   * @internal
   */
  className: string;
  /**
   * @internal
   */
  edmxPath: PathLike;
}
/**
 * @internal
 */
export interface VdmServiceEntities {
  /**
   * @internal
   */
  entities: VdmEntity[];
  /**
   * @internal
   */
  complexTypes: VdmComplexType[];
  /**
   * @internal
   */
  enumTypes: VdmEnumType[];
  /**
   * @internal
   */
  functionImports: VdmOperation[];
  /**
   * @internal
   */
  actionImports?: VdmOperation[];
}
/**
 * @internal
 */
export type VdmServiceMetadata = VdmServicePackageMetaData & VdmServiceEntities;

/**
 * Entity
 * @internal
 */
export interface VdmEntity {
  // todo uniqueness check like request builder
  // apiName: string;
  /**
   * @internal
   */
  entitySetName: string;
  /**
   * @internal
   */
  entityTypeName: string;
  /**
   * @internal
   */
  className: string;
  /**
   * @internal
   */
  creatable: boolean;
  /**
   * @internal
   */
  updatable: boolean;
  /**
   * @internal
   */
  deletable: boolean;
  /**
   * @internal
   */
  keys: VdmProperty[];
  /**
   * @internal
   */
  properties: VdmProperty[];
  /**
   * @internal
   */
  navigationProperties: VdmNavigationProperty[];
  /**
   * @internal
   */
  description: string;
  /**
   * @internal
   */
  entityTypeNamespace: string;
  /**
   * @internal
   */
  functions: VdmOperation[];
  /**
   * @internal
   */
  actions: VdmOperation[];
}

/**
 * Subset of Entity required for generating bound actions and functions.
 * This is required because at the time of generating bound actions and functions the full list of entities is not yet available.
 * But the properties included here are needed for generating the functions and actions.
 * @internal
 */
export type VdmPartialEntity = Pick<
  VdmEntity,
  'entityTypeName' | 'entityTypeNamespace' | 'className'
>;

/**
 * Properties
 * @internal
 */
export interface VdmPropertyDescriptor {
  /**
   * @internal
   */
  originalName: string;
  /**
   * @internal
   */
  instancePropertyName: string;
  /**
   * @internal
   */
  propertyNameAsParam: string;
  /**
   * @internal
   */
  staticPropertyName: string;
  /**
   * @internal
   */
  isComplex?: boolean;
  /**
   * @internal
   */
  isEnum?: boolean;
  /**
   * @internal
   */
  isCollection: boolean;
}
/**
 * @internal
 */
export interface VdmPropertyValueConstraints {
  /**
   * @internal
   */
  maxLength?: string;
  /**
   * @internal
   */
  nullable: boolean;
}
/**
 * @internal
 */
export interface VdmProperty
  extends VdmPropertyDescriptor,
    VdmPropertyValueConstraints,
    VdmMappedEdmType {
  /**
   * @internal
   */
  description: string;
}

/**
 * Navigation Properties
 * @internal
 */
export interface VdmNavigationProperty extends VdmPropertyDescriptor {
  /**
   * @internal
   */
  from: string;
  /**
   * @internal
   */
  to: string;
  /**
   * @internal
   */
  toEntityClassName: string;
}
/**
 * @internal
 */
export interface VdmEnumMemberType {
  /**
   * @internal
   */
  name: string;
  /**
   * @internal
   */
  originalValue: string;
}
/**
 * @internal
 */
export interface VdmEnumType {
  /**
   * @internal
   */
  originalName: string;
  /**
   * @internal
   */
  typeName: string;
  /**
   * @internal
   */
  underlyingType: string;
  /**
   * @internal
   */
  members: VdmEnumMemberType[];
}

/**
 * Complex types and function imports
 * @internal
 */
export interface VdmComplexType {
  /**
   * @internal
   */
  originalName: string;
  /**
   * @internal
   */
  properties: VdmProperty[];
  /**
   * @internal
   */
  typeName: string;
  /**
   * @internal
   */
  fieldType: string;
  /**
   * @internal
   */
  namespace: string;
}
/**
 * @internal
 * Represents a partial operation (action or function).
 */
export interface VdmOperationBase {
  /**
   * @internal
   */
  originalName: string;
  /**
   * @internal
   */
  parameters: VdmParameter[];
  /**
   * @internal
   */
  parametersTypeName: string;
  /**
   * @internal
   */
  name: string;
  /**
   * @internal
   */
  description: string;
  /**
   * @internal
   */
  type: 'function' | 'action';
  /**
   * @internal
   */
  isBound: boolean;
  /**
   * @internal
   * only set for bound operations
   */
  entityClassName?: string;
}

/**
 * @internal
 * Represents either a function or an action.
 */
export interface VdmOperation extends VdmOperationBase {
  /**
   * @internal
   */
  httpMethod: string;
  /**
   * @internal
   */
  returnType: VdmOperationReturnType;
}

/**
 * @internal
 */
export interface VdmOperationReturnType {
  /**
   * @internal
   */
  builderFunction?: string;
  /**
   * @internal
   */
  returnType: string;
  /**
   * @internal
   */
  isCollection: boolean;
  /**
   * @internal
   */
  isNullable: boolean;
  /**
   * @internal
   */
  returnTypeCategory: VdmReturnTypeCategory;
  /**
   * @internal
   */
  unsupportedReason?: VdmUnsupportedReason;
}
/**
 * @internal
 */
export enum VdmReturnTypeCategory {
  ENTITY,
  COMPLEX_TYPE,
  EDM_TYPE,
  VOID,
  NEVER
}
/**
 * @internal
 */
export enum VdmUnsupportedReason {
  ENTITY_NOT_DESERIALIZABLE
}
/**
 * @internal
 */
export interface VdmParameter extends VdmMappedEdmType {
  /**
   * @internal
   */
  originalName: string;
  /**
   * @internal
   */
  parameterName: string;
  /**
   * @internal
   */
  nullable: boolean;
  /**
   * @internal
   */
  description: string;
}
/**
 * @internal
 */
export interface VdmMappedEdmType {
  /**
   * @internal
   */
  edmType: string;
  /**
   * @internal
   */
  jsType: string;
  /**
   * @internal
   */
  fieldType: string;
}
/**
 * @internal
 */
export interface ApiBusinessHubMetadata {
  /**
   * @internal
   */
  url: string;
  /**
   * @internal
   */
  communicationScenario: string | null;
  /**
   * @internal
   */
  businessDocumentationUrl?: string;
}
