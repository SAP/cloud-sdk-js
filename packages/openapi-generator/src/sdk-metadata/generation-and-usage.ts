import { MultiLineText } from '@sap-cloud-sdk/generator-common/internal';
import { OpenApiDocument } from '../openapi-types';
import { apiSpecificCodeSample } from './code-sample';
import { getMainApi, getMainOperation } from './code-sample-util';

/**
 * @internal
 */
export function getApiSpecificUsage(
  openApiDocument: OpenApiDocument
): MultiLineText {
  const apisWithOperations = openApiDocument.apis.filter(
    api => api.operations?.length > 0
  );

  if (apisWithOperations.length > 0) {
    const mainApi = getMainApi(openApiDocument.serviceName, apisWithOperations);
    const operation = getMainOperation(mainApi);

    const instructions = apiSpecificCodeSample(
      mainApi.name,
      operation,
      openApiDocument.serviceOptions.directoryName
    );
    return instructions;
  }
  return '';
}
