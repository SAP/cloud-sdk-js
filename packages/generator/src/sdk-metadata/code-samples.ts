import { codeBlock } from '@sap-cloud-sdk/util';
import { MultiLineText } from '@sap-cloud-sdk/generator-common/internal';
import voca from 'voca';
import { VdmOperation } from '../vdm-types';
import { getApiName } from '../generator-without-ts-morph/service';
import { getOperationParams } from './code-sample-util';

/**
 * @internal
 */
export function entityCodeSample(
  entityName: string,
  serviceName: string,
  directoryName: string
): MultiLineText {
  return codeBlock`
import { ${voca.decapitalize(
    serviceName
  )} } from './generated/${directoryName}';

const { ${getApiName(entityName)} } = ${voca.decapitalize(serviceName)}();
const resultPromise = ${getApiName(
    entityName
  )}.requestBuilder().getAll().top(5).execute({ destinationName: 'myDestinationName' });

`;
}

/**
 * @internal
 */
export function functionImportCodeSample(
  functionImport: VdmOperation,
  packageName: string
): MultiLineText {
  return importsCodeSample(functionImport, packageName);
}
/**
 * @internal
 */
export function actionImportCodeSample(
  actionImport: VdmOperation,
  packageName: string
): MultiLineText {
  return importsCodeSample(actionImport, packageName);
}

function importsCodeSample(
  actionFunctionImport: VdmOperation,
  packageName: string
): MultiLineText {
  return codeBlock`
import { ${actionFunctionImport.name}${
    actionFunctionImport.parametersTypeName
      ? `, ${actionFunctionImport.parametersTypeName}`
      : ''
  } } from './generated/${packageName}';

${getParameterCodeSample(actionFunctionImport)}
const resultPromise = ${actionFunctionImport.name}(${
    actionFunctionImport.parametersTypeName ? 'parameter' : ''
  }).execute({ destinationName: 'myDestinationName' });
`;
}

function getParameterCodeSample(functionImport: VdmOperation): string {
  if (functionImport.parameters) {
    return `const parameter: ${
      functionImport.parametersTypeName
    } = ${getOperationParams(functionImport.parameters)};
    `;
  }
  return '';
}
