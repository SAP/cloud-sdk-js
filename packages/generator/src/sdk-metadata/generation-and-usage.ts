import { MultiLineText } from '@sap-cloud-sdk/generator-common/internal';
import { VdmServiceMetadata } from '../vdm-types';
import {
  actionImportCodeSample,
  entityCodeSample,
  functionImportCodeSample
} from './code-samples';
import { sampleOperationImport, getODataEntity } from './code-sample-util';

/**
 * @internal
 */
export function getApiSpecificUsage(
  service: VdmServiceMetadata
): MultiLineText {
  if (service.entities?.length > 0) {
    const entity = getODataEntity(service.originalFileName, service.entities);
    return entityCodeSample(
      entity.className,
      service.className,
      service.directoryName
    );
  }
  // Return function/action import usage if no entity is found.
  if (service.functionImports?.length > 0) {
    const functionImport = sampleOperationImport(
      service.originalFileName,
      service.functionImports
    );
    return functionImportCodeSample(
      functionImport,
      `${service.directoryName}/function-imports`
    );
  }
  if (service.actionImports) {
    const actionImport = sampleOperationImport(
      service.originalFileName,
      service.actionImports
    );
    return actionImportCodeSample(
      actionImport,
      `${service.directoryName}/action-imports`
    );
  }
  return '';
}
