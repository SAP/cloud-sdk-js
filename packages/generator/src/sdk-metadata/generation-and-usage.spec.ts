import { resolve } from 'path';
import { writeFile, readFile, removeSync } from 'fs-extra';
import execa from 'execa';
import { VdmServiceMetadata } from '../vdm-types';
import { getApiSpecificUsage } from './generation-and-usage';
import { entityCodeSample } from './code-samples';

describe('generation-and-usage', () => {
  const service = {
    npmPackageName: '@sap/dummy-package',
    originalFileName: 'DummyClass',
    directoryName: 'dummy-service',
    className: 'dummyService',
    entities: [{ className: 'DummyCollection' }, { className: 'DummyClass' }]
  } as VdmServiceMetadata;

  const serviceWithFunctionImport = {
    npmPackageName: '@sap/dummy-package',
    originalFileName: 'Dummy',
    functionImports: [
      {
        name: 'dummyFunc',
        parametersTypeName: 'dummyParamType',
        parameters: [
          { parameterName: 'dummyParam1' },
          { parameterName: 'dummyParam2' },
          { parameterName: 'dummyParam3' }
        ]
      }
    ]
  } as VdmServiceMetadata;

  const serviceWithMultipleImports = {
    npmPackageName: '@sap/dummy-package',
    originalFileName: 'Dummy',
    functionImports: [
      {
        name: 'dummyFunc',
        httpMethod: 'get',
        parametersTypeName: 'dummyParamType',
        parameters: [{ parameterName: 'dummyParam1' }]
      },
      {
        name: 'dummyFunction',
        httpMethod: 'get',
        parameters: [
          { parameterName: 'dummyParam2' },
          { parameterName: 'dummyParam3' }
        ]
      }
    ]
  } as VdmServiceMetadata;

  const serviceWithActionImport = {
    npmPackageName: '@sap/dummy-package',
    originalFileName: 'Dummy',
    actionImports: [
      {
        name: 'dummyActionImport',
        httpMethod: 'get',
        parametersTypeName: 'dummyParamType',
        parameters: [{ parameterName: 'dummyParam1' }]
      }
    ]
  } as VdmServiceMetadata;

  it('creates api specific usage for entity', () => {
    expect(getApiSpecificUsage(service)).toMatchSnapshot();
  });

  it('create api specific usage for function import', () => {
    expect(getApiSpecificUsage(serviceWithFunctionImport)).toMatchSnapshot();
  });

  it('create api specific usage for function import with minimum parameters', () => {
    expect(getApiSpecificUsage(serviceWithMultipleImports)).toMatchSnapshot();
  });

  it('create api specific usage for action import', () => {
    expect(getApiSpecificUsage(serviceWithActionImport)).toMatchSnapshot();
  });

  xit('creates compiling generic usage', async () => {
    const codeSnippet = entityCodeSample(
      'TestEntity',
      'TestService',
      '@sap-cloud-sdk/test-services-odata-v2/test-service'
    );
    const tsFile = 'generic-get-all-code-sample.ts';
    const jsFile = tsFile.replace('.ts', '.js');
    await writeFile(resolve(__dirname, tsFile), codeSnippet);
    await execa('npx', [
      'tsc',
      resolve(__dirname, tsFile),
      '--esModuleInterop',
      '--target',
      'es2019',
      '--module',
      'commonjs'
    ]);
    await expect(readFile(resolve(__dirname, jsFile))).resolves.toBeDefined();
    [tsFile, jsFile].map(file => removeSync(resolve(__dirname, file)));
  }, 60000);
});
