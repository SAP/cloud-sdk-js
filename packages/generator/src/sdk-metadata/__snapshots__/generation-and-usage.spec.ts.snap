// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generation-and-usage create api specific usage for action import 1`] = `
"import { dummyActionImport, dummyParamType } from './generated/undefined/action-imports';

const parameter: dummyParamType = { dummyParam1: 'dummyParam1' };
    
const resultPromise = dummyActionImport(parameter).execute({ destinationName: 'myDestinationName' });"
`;

exports[`generation-and-usage create api specific usage for function import 1`] = `
"import { dummyFunc, dummyParamType } from './generated/undefined/function-imports';

const parameter: dummyParamType = { dummyParam1: 'dummyParam1', dummyParam2: 'dummyParam2', ... };
    
const resultPromise = dummyFunc(parameter).execute({ destinationName: 'myDestinationName' });"
`;

exports[`generation-and-usage create api specific usage for function import with minimum parameters 1`] = `
"import { dummyFunc, dummyParamType } from './generated/undefined/function-imports';

const parameter: dummyParamType = { dummyParam1: 'dummyParam1' };
    
const resultPromise = dummyFunc(parameter).execute({ destinationName: 'myDestinationName' });"
`;

exports[`generation-and-usage creates api specific usage for entity 1`] = `
"import { dummyService } from './generated/dummy-service';

const { dummyClassApi } = dummyService();
const resultPromise = dummyClassApi.requestBuilder().getAll().top(5).execute({ destinationName: 'myDestinationName' });
"
`;
